doc/topics/the-event-loop-timers-and-nexttick.md

```diff
@@ -0,0 +1,358 @@
+remaining until the soonest timer's threshold is reached.  While it is
+waiting 95 ms pass, `fs.readFile()` finishes reading the file and its
+callback which takes 10 ms to complete is added to the `poll` queue and
+executed.  When the callback finishes, there are no more callbacks in the
+queue, so the event loop will see  that the threshold of the soonest timer has
+been reached then wrap back to the `timers` phase to execute the  timer's
+callback.  In this example, you will see that the total delay between
+the timer being scheduled and its callback being executed will be
+105ms.
+
+Note: To prevent the `poll` phase from starving the event loop, libuv also has
+a hard maximum (system dependent) before it stops poll'ing for more events.
+
+### pending callbacks:
+
+This phase executes callbacks for some system operations such as types of TCP
+errors.  For example if a TCP socket receives `ECONNREFUSED` when
+attempting to connect, some *nix systems want to wait to report the
+error. This will be queued to execute in the `pending callbacks` phase.
+
+### poll:  
+
+When the event loop enters the `poll` phase _and there are no timers
+scheduled_, one of two things will happend: 
+
+* _if the `poll` queue **is not empty**_, the event loop will iterate through 
+its queue of callbacks executing them synchronously until either the queue has 
+been exhausted, or the system-dependent hard limit is reached. 
+
+* _if the `poll` queue is **empty**_, the event loop will wait for a
+callback
+to be added to the queue, then execute it immediately. 
+
+Once the `poll` queue is empty the event loop will check for timers _whose
+time thresholds have been reached_.  If one or more timers are ready, the
+event loop will wrap back to the timers phase to execute  those timers'
+callbacks.
+
+### `setImmediate()`:
+
+`setImmediate()` is actually a special timer that runs in a separate phase of
+the event loop.  It uses a libuv API that schedules callbacks to execute after
+the `poll` phase has completed.
+
+Generally, as the code is executed, the event loop will eventually hit the
+`poll` phase where it will wait for an incoming connection, request, etc.
+However, after a callback has been scheduled with `setImmediate()`,
+then the `poll` phase becomes idle, it will end and continue to the
+`setImmediate` phase rather than waiting for `poll` events. 
+
+## `setImmediate()` vs `setTimeout()`
+
+How quickly a `setImmediate()` callback is executed is only limited by how
+quickly the event loop can be processed whereas a timer won't fire until the
+number of milliseconds passed have elapsed.
+
+The advantage to using `setImmediate()` over `setTimeout()` is that the lowest
+value you may set a timer's delay to is 1 ms (0 is coerced to 1) which doesn't
+seem like much time to us humans, but it's actually pretty slow compared to
+how quickly `setImmediate()` can execute — the event loop operates on the
+microsecond scale (1 ms = 1000 µs.)
```

Comment:
The `setImmediate` list is executed once per loop iteration, right after polling for i/o (source https://github.com/nodejs/node/blob/master/src/node.cc#L2724-L2747 and http://docs.libuv.org/en/v1.x/check.html).

There are in fact two "timers" phases in libuv https://github.com/libuv/libuv/blob/c0fa2e7518a3a0e364c56f8223bdd0f549ddac66/src/unix/core.c#L331-L380, because node runs `uv_run` with `UV_RUN_ONCE`:
https://github.com/nodejs/node/blob/master/src/node.cc#L4170.

Your examples are not happening within a i/o cycle, so you are bound to _how fast your process go_, which is impacted by other programs running on your machine. 
If you move the two calls within a i/o cycle, the immediate callback is always executed first:

```js
var fs = require('fs')

fs.readFile(__filename, () => {
  setTimeout(() => {
    console.log('timeout')
  }, 0)
  setImmediate(() => {
    console.log('immediate')
  })
})
```

@saghul @cjihrig for confirmation.